Python/ceval.c

```diff
@@ -4617,6 +4635,40 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             DISPATCH();
         }
 
+        TARGET(PRECALL_PY_CLASS) {
+            SpecializedCacheEntry *cache = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &cache[0].adaptive;
+            _PyCallCache *cache1 = &cache[-1].call;
+            int original_oparg = cache->adaptive.original_oparg;
+            int is_method = (PEEK(original_oparg + 2) != NULL);
+            DEOPT_IF(is_method, PRECALL);
+            PyObject *cls = PEEK(original_oparg + 1);
+            DEOPT_IF(!PyType_Check(cls), PRECALL);
+            PyTypeObject *cls_t = (PyTypeObject *)cls;
+            DEOPT_IF(cls_t->tp_version_tag != cache0->version, PRECALL);
+            assert(cls_t->tp_flags & Py_TPFLAGS_HEAPTYPE);
+            PyObject *init = ((PyHeapTypeObject *)cls_t)->_spec_cache.init;
+            assert(PyFunction_Check(init));
+            DEOPT_IF(((PyFunctionObject *)(init))->func_version != cache1->func_version, PRECALL);
+            DEOPT_IF(cls_t->tp_new != PyBaseObject_Type.tp_new, PRECALL);
+            STAT_INC(PRECALL, hit);
+
+            PyObject *self = _PyObject_New_Vector(cls_t, &PEEK(original_oparg),
```

Comment:
If the specializer only specialized for classes that don't override `__new__`, you can avoid calling `__new__` and just construct the object directly.

Reply:
`_PyObject_New_Vector` does some argument checking. https://github.com/python/cpython/pull/31707/files#diff-1decebeef15f4e0b0ce106c665751ec55068d4d1d1825847925ad4f528b5b872R4525

Come to think of it, if we verify that the argcount is right at specialization time, do we need to re-verify at runtime? Would it be safe to call `tp_alloc` directly? It seems that the only thing that could change is the `kwds` dict, but even then  that's only used for argument count checking again.